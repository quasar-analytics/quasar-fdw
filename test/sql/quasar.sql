/* Select */
/* Basic selection with limit */
SELECT * FROM smallzips ORDER BY city LIMIT 3;
/* Select less fields than exist */
SELECT city FROM smallzips ORDER BY city LIMIT 1;
/* Basic WHERE clause */
SELECT * FROM zips WHERE "state" = 'CO' ORDER BY city LIMIT 2;
/* Nested selection */
SELECT * FROM nested LIMIT 1;
/* less fields than in relation, with one in a WHERE clause */
SELECT city FROM zips WHERE "state" = 'CO' ORDER BY city LIMIT 1;
SELECT city,pop FROM smallzips WHERE pop % 2 = 1 ORDER BY city LIMIT 3;
/* Test out array usage */
SELECT * FROM zipsloc ORDER BY loc[1] LIMIT 2;
/* Test out json usage */
SELECT loc->0 AS loc0, locb->1 AS loc1, locb FROM zipsjson ORDER BY loc->0 LIMIT 2;
/* Pushdown regex operators */
SELECT * FROM zips WHERE "state" LIKE 'A%' ORDER BY city LIMIT 3;
SELECT * FROM smallzips WHERE "city" !~~ 'B%' ORDER BY city LIMIT 3;
/* pushdown math operators */
SELECT * FROM smallzips WHERE pop > 1000 AND pop + pop <= 10000 ORDER BY city LIMIT 3;
/* join zips and zipsjson */
SELECT zips.city AS city, pop, state, loc
       FROM smallzips zips JOIN zipsjson ON zips.city = zipsjson.city
       ORDER BY zips.city LIMIT 3;
/* query for a missing field */
SELECT missing, city FROM zips_missing ORDER BY city LIMIT 3;
/* query with a bad type */
SELECT loc FROM zips_bad LIMIT 1;
/* Make quasar_fdw convert a float to integer
 * Quasar converts pop to a float and we convert it back */
SELECT pop, loc0 FROM zips_convert WHERE loc0 < -70 ORDER BY pop LIMIT 3;
/* convert string timestamps to timestamp values */
SELECT * FROM commits ORDER BY ts LIMIT 5;
/* BUT we shouldn't push down comparisons on dates because the underlying data is string */
SELECT * FROM commits WHERE ts = timestamp 'Thu Jan 29 15:52:37 2015';
/* convert year as string to integers floats and dates */
SELECT yr, yearint, yearfloat FROM olympics WHERE "yr" = '1924' LIMIT 3;
/* pushdown of concat */
SELECT * FROM smallzips WHERE length(concat(state, city)) > 4 /* push down concat columns */
                        AND state = concat('M'::char, 'A'::char) /* pushed down correctly */ ORDER BY city LIMIT 5;
/* LIKE operator only supports constant right sides */
SELECT * FROM smallzips WHERE state LIKE concat('B'::char, '%'::char) ORDER BY city;
/* array expansion */
SELECT * FROM user_comments ORDER BY comment_text;
SELECT * FROM nested_expansion ORDER BY vals;
/* ORDER BY pushdown */
SELECT * FROM zips ORDER BY length(city), -pop, state LIMIT 10;
/* Date pushdown and ordering */
SELECT * FROM commits_timestamps WHERE ts < TIMESTAMP '2015-01-20T00:00:00Z' ORDER BY ts LIMIT 2;
SELECT * FROM commits_timestamps WHERE ts < DATE '2015-01-20' ORDER BY ts LIMIT 2;
SELECT * FROM commits_timestamps WHERE tstz < TIMESTAMPTZ '2015-01-15 19:43:04 PST' ORDER BY ts;
SELECT * FROM commits_timestamps ORDER BY ts DESC LIMIT 2;
/* Bad types should emit error contexts appropriately */
SELECT city FROM zips_badtype ORDER BY city;
SELECT state FROM zips_badtype ORDER BY state;
SELECT pop FROM zips_badtype ORDER BY pop;
/* Aggregations */
SELECT count(*) FROM smallzips;
SELECT count(*) FROM zips WHERE state IN ('CA','OR','WA');
/* Array subscripts push down correctly from 1-based to 0-based */
SELECT loc FROM zipsloc WHERE loc[1] < 0 ORDER BY loc[1] LIMIT 3;
SELECT loc FROM zipsloc WHERE loc[1+1] > 0 ORDER BY loc[2] LIMIT 3;
/* Scalar Array ops */
SELECT * FROM smallzips WHERE state IN ('MA', 'CA') ORDER BY city,pop LIMIT 3;
SELECT * FROM smallzips WHERE state IN ('MA') ORDER BY city,pop LIMIT 3;
