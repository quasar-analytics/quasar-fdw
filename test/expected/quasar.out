/* Select */
/* Basic selection with limit */
SELECT * FROM smallzips ORDER BY city LIMIT 3;
   city   |  pop  | state 
----------+-------+-------
 ADAMS    |  9901 | MA
 AGAWAM   | 15338 | MA
 ASHFIELD |  1535 | MA
(3 rows)

/* Select less fields than exist */
SELECT city FROM smallzips ORDER BY city LIMIT 1;
 city  
-------
 ADAMS
(1 row)

/* Basic WHERE clause */
SELECT * FROM zips WHERE "state" = 'CO' ORDER BY city LIMIT 2;
  city   | pop | state 
---------+-----+-------
 AGATE   | 230 | CO
 AGUILAR | 928 | CO
(2 rows)

/* Nested selection */
SELECT * FROM nested LIMIT 1;
 a | b | c 
---+---+---
 m | n | o
(1 row)

/* less fields than in relation, with one in a WHERE clause */
SELECT city FROM zips WHERE "state" = 'CO' ORDER BY city LIMIT 1;
 city  
-------
 AGATE
(1 row)

SELECT city,pop FROM smallzips WHERE pop % 2 = 1 ORDER BY city LIMIT 3;
     city     | pop  
--------------+------
 ADAMS        | 9901
 ASHFIELD     | 1535
 ASHLEY FALLS |  561
(3 rows)

/* Test out array usage */
SELECT * FROM zipsloc ORDER BY loc[1] LIMIT 2;
          loc           
------------------------
 {-73.456575,42.101153}
 {-73.38251,42.334752}
(2 rows)

/* Test out json usage */
SELECT loc->0 AS loc0, locb->1 AS loc1, locb FROM zipsjson ORDER BY city, pop LIMIT 2;
    loc0    |   loc1    |          locb           
------------+-----------+-------------------------
 -73.117225 | 42.622319 | [-73.117225, 42.622319]
 -72.622739 | 42.070206 | [-72.622739, 42.070206]
(2 rows)

/* Pushdown regex operators */
SELECT * FROM zips WHERE "state" LIKE 'A%' ORDER BY city LIMIT 3;
    city    |  pop  | state 
------------+-------+-------
 ABBEVILLE  |  5416 | AL
 ACMAR      |  6055 | AL
 ADAMSVILLE | 10616 | AL
(3 rows)

SELECT * FROM smallzips WHERE "city" !~~ 'B%' ORDER BY city LIMIT 3;
   city   |  pop  | state 
----------+-------+-------
 ADAMS    |  9901 | MA
 AGAWAM   | 15338 | MA
 ASHFIELD |  1535 | MA
(3 rows)

/* pushdown math operators */
SELECT * FROM smallzips WHERE pop > 1000 AND pop + pop <= 10000 ORDER BY city LIMIT 3;
   city   | pop  | state 
----------+------+-------
 ASHFIELD | 1535 | MA
 BARRE    | 4546 | MA
 BECKET   | 1070 | MA
(3 rows)

/* join zips and zipsjson */
SELECT zips.city AS city, pop, state, loc
       FROM smallzips zips JOIN zipsjson ON zips.city = zipsjson.city
       ORDER BY zips.city LIMIT 3;
ERROR:  column reference "pop" is ambiguous
LINE 2: SELECT zips.city AS city, pop, state, loc
                                  ^
/* query for a missing field */
SELECT missing, city FROM zips_missing ORDER BY city LIMIT 3;
 missing |   city   
---------+----------
         | ADAMS
         | AGAWAM
         | ASHFIELD
(3 rows)

/* query with a bad type */
SELECT loc FROM zips_bad LIMIT 1;
WARNING:  quasar_fdw: column loc is scalar type but got json/array response.
 loc 
-----
    
(1 row)

/* Make quasar_fdw convert a float to integer
 * Quasar converts pop to a float and we convert it back */
SELECT pop, loc0 FROM zips_convert WHERE loc0 < -70 ORDER BY pop LIMIT 3;
 pop |    loc0    
-----+------------
  16 | -72.764124
  97 | -72.960156
 122 | -72.844092
(3 rows)

/* convert string timestamps to timestamp values */
SELECT * FROM commits ORDER BY ts LIMIT 5;
            ts            |                   sha                    |   author_name   |     author_email      |                                                url                                                | comment_count 
--------------------------+------------------------------------------+-----------------+-----------------------+---------------------------------------------------------------------------------------------------+---------------
 Wed Jan 14 23:10:11 2015 | 43b4018ae5d1e40bbbc2babb8929ed247b5d2dcb | Moss Prescott   | moss@theprescotts.com | https://api.github.com/repos/slamdata/slamengine/commits/43b4018ae5d1e40bbbc2babb8929ed247b5d2dcb |             0
 Thu Jan 15 20:43:04 2015 | 3de46760d45c4fd0db9a03ad978361df3e8f0998 | Moss Prescott   | moss@theprescotts.com | https://api.github.com/repos/slamdata/slamengine/commits/3de46760d45c4fd0db9a03ad978361df3e8f0998 |             0
 Fri Jan 16 02:08:26 2015 | 6a23a4c8809635828bbdeea145894287348c9bd5 | Moss Prescott   | moss@theprescotts.com | https://api.github.com/repos/slamdata/slamengine/commits/6a23a4c8809635828bbdeea145894287348c9bd5 |             0
 Fri Jan 16 15:46:22 2015 | 5b54522e340244d618645ace4bd0cbb7edf8bd5b | John A. De Goes | john@degoes.net       | https://api.github.com/repos/slamdata/slamengine/commits/5b54522e340244d618645ace4bd0cbb7edf8bd5b |             0
 Fri Jan 16 16:54:56 2015 | ad8a6a73f898860b48f3d71ed6110b3506a8c898 | Moss Prescott   | moss@theprescotts.com | https://api.github.com/repos/slamdata/slamengine/commits/ad8a6a73f898860b48f3d71ed6110b3506a8c898 |             0
(5 rows)

/* BUT we shouldn't push down comparisons on dates because the underlying data is string */
SELECT * FROM commits WHERE ts = timestamp 'Thu Jan 29 15:52:37 2015';
            ts            |                   sha                    | author_name |     author_email     |                                                url                                                | comment_count 
--------------------------+------------------------------------------+-------------+----------------------+---------------------------------------------------------------------------------------------------+---------------
 Thu Jan 29 15:52:37 2015 | b8a2302e6a0659875d03bfe4988c000f2ed027a0 | Greg Pfeil  | greg@technomadic.org | https://api.github.com/repos/slamdata/slamengine/commits/b8a2302e6a0659875d03bfe4988c000f2ed027a0 |             0
(1 row)

/* convert year as string to integers floats and dates */
SELECT yr, yearint, yearfloat FROM olympics WHERE "yr" = '1924' LIMIT 3;
  yr  | yearint | yearfloat 
------+---------+-----------
 1924 |    1924 |      1924
 1924 |    1924 |      1924
 1924 |    1924 |      1924
(3 rows)

/* pushdown of concat */
SELECT * FROM smallzips WHERE length(concat(state, city)) > 4 /* push down concat columns */
                        AND state = concat('M'::char, 'A'::char) /* pushed down correctly */ ORDER BY city LIMIT 5;
     city     |  pop  | state 
--------------+-------+-------
 ADAMS        |  9901 | MA
 AGAWAM       | 15338 | MA
 ASHFIELD     |  1535 | MA
 ASHLEY FALLS |   561 | MA
 BARRE        |  4546 | MA
(5 rows)

/* LIKE operator only supports constant right sides */
SELECT * FROM smallzips WHERE state LIKE concat('B'::char, '%'::char) ORDER BY city;
 city | pop | state 
------+-----+-------
(0 rows)

/* array expansion */
SELECT * FROM user_comments ORDER BY comment_text;
 user_id | profile_name | age | title | comment_id | comment_text | comment_reply_to_profile | comment_reply_to_comment | comment_time 
---------+--------------+-----+-------+------------+--------------+--------------------------+--------------------------+--------------
 8927524 | Mary Jane    |  29 | Dr    | GH732AFC   |              |                  9654726 | A44124F                  | 03-01-2015
 8927524 | Mary Jane    |  29 | Dr    | F2372BAC   | I concur.    |                  9817361 | F8ACD164F                | 02-03-2015
(2 rows)

SELECT * FROM nested_expansion ORDER BY vals;
 vals |                                                       topobj                                                        
------+---------------------------------------------------------------------------------------------------------------------
    7 | {"midArr":[[10,11,12],{"a":"j","b":"k","c":"l"}],"midObj":{"botArr":[13,14,15],"botObj":{"a":"m","b":"n","c":"o"}}}
    8 | {"midArr":[[10,11,12],{"a":"j","b":"k","c":"l"}],"midObj":{"botArr":[13,14,15],"botObj":{"a":"m","b":"n","c":"o"}}}
    9 | {"midArr":[[10,11,12],{"a":"j","b":"k","c":"l"}],"midObj":{"botArr":[13,14,15],"botObj":{"a":"m","b":"n","c":"o"}}}
(3 rows)

/* ORDER BY pushdown */
SELECT * FROM zips ORDER BY length(city), -pop, state LIMIT 10;
 city |  pop  | state 
------+-------+-------
 M M  | 32901 | GA
 ADA  | 24967 | OK
 ROY  | 22166 | UT
 ORA  | 21437 | SC
 RYE  | 16479 | NY
 LEE  | 15487 | NH
 ORA  |  9987 | SC
 OPP  |  9901 | AL
 ADA  |  9568 | MI
 YOE  |  8691 | PA
(10 rows)

/* Date pushdown and ordering */
SELECT * FROM commits_timestamps WHERE ts < TIMESTAMP '2015-01-20T00:00:00Z' ORDER BY ts LIMIT 2;
            ts            |             tstz             |                   sha                    
--------------------------+------------------------------+------------------------------------------
 Wed Jan 14 23:10:11 2015 | Wed Jan 14 15:10:11 2015 PST | 43b4018ae5d1e40bbbc2babb8929ed247b5d2dcb
 Thu Jan 15 20:43:04 2015 | Thu Jan 15 12:43:04 2015 PST | 3de46760d45c4fd0db9a03ad978361df3e8f0998
(2 rows)

SELECT * FROM commits_timestamps WHERE ts < DATE '2015-01-20' ORDER BY ts LIMIT 2;
            ts            |             tstz             |                   sha                    
--------------------------+------------------------------+------------------------------------------
 Wed Jan 14 23:10:11 2015 | Wed Jan 14 15:10:11 2015 PST | 43b4018ae5d1e40bbbc2babb8929ed247b5d2dcb
 Thu Jan 15 20:43:04 2015 | Thu Jan 15 12:43:04 2015 PST | 3de46760d45c4fd0db9a03ad978361df3e8f0998
(2 rows)

SELECT * FROM commits_timestamps WHERE tstz < TIMESTAMPTZ '2015-01-15 19:43:04 PST' ORDER BY ts;
            ts            |             tstz             |                   sha                    
--------------------------+------------------------------+------------------------------------------
 Wed Jan 14 23:10:11 2015 | Wed Jan 14 15:10:11 2015 PST | 43b4018ae5d1e40bbbc2babb8929ed247b5d2dcb
 Thu Jan 15 20:43:04 2015 | Thu Jan 15 12:43:04 2015 PST | 3de46760d45c4fd0db9a03ad978361df3e8f0998
 Fri Jan 16 02:08:26 2015 | Thu Jan 15 18:08:26 2015 PST | 6a23a4c8809635828bbdeea145894287348c9bd5
(3 rows)

SELECT * FROM commits_timestamps ORDER BY ts DESC LIMIT 2;
            ts            |             tstz             |                   sha                    
--------------------------+------------------------------+------------------------------------------
 Thu Jan 29 15:52:37 2015 | Thu Jan 29 07:52:37 2015 PST | b8a2302e6a0659875d03bfe4988c000f2ed027a0
 Thu Jan 29 00:23:14 2015 | Wed Jan 28 16:23:14 2015 PST | 18164870e8803e5885a88f8dfd53b82c80721503
(2 rows)

/* Bad types should emit error contexts appropriately */
SELECT city FROM zips_badtype ORDER BY city;
ERROR:  invalid input syntax for integer: "AARON"
CONTEXT:  column "city" of foreign table "zips_badtype"
SELECT state FROM zips_badtype ORDER BY state;
ERROR:  value too long for type character(1)
CONTEXT:  column "state" of foreign table "zips_badtype"
SELECT pop FROM zips_badtype ORDER BY pop;
ERROR:  date/time field value out of range: "0"
HINT:  Perhaps you need a different "datestyle" setting.
CONTEXT:  column "pop" of foreign table "zips_badtype"
/* Aggregations */
SELECT count(state) FROM smallzips;
 count 
-------
   100
(1 row)

SELECT count(state) FROM zips WHERE state IN ('CA','OR','WA');
 count 
-------
  2384
(1 row)

/* Array subscripts push down correctly from 1-based to 0-based */
SELECT loc FROM zipsloc WHERE loc[1] < 0 ORDER BY loc[1] LIMIT 3;
          loc           
------------------------
 {-73.456575,42.101153}
 {-73.38251,42.334752}
 {-73.374544,42.265296}
(3 rows)

SELECT loc FROM zipsloc WHERE loc[1+1] > 0 ORDER BY loc[2] LIMIT 3;
          loc           
------------------------
 {-72.5676,42.050658}
 {-72.770588,42.051099}
 {-73.320195,42.059552}
(3 rows)

/* Scalar Array ops */
SELECT * FROM smallzips WHERE state IN ('MA', 'CA') ORDER BY city,pop LIMIT 3;
   city   |  pop  | state 
----------+-------+-------
 ADAMS    |  9901 | MA
 AGAWAM   | 15338 | MA
 ASHFIELD |  1535 | MA
(3 rows)

SELECT * FROM smallzips WHERE state IN ('MA') ORDER BY city,pop LIMIT 3;
   city   |  pop  | state 
----------+-------+-------
 ADAMS    |  9901 | MA
 AGAWAM   | 15338 | MA
 ASHFIELD |  1535 | MA
(3 rows)

/* Spaced fields */
SELECT * FROM weird_fields;
 spaced | underscored |                       toparray                        
--------+-------------+-------------------------------------------------------
        |             | [[4,5,6],{"a":"d","b":"e","c":"f"}]
        |             | {"botArr":[7,8,9],"botObj":{"a":"g","b":"h","c":"i"}}
(2 rows)

